### PR TITLE
Experiment: manual jupyter server connection through uri server provider api

### PR DESCRIPTION
**Pull request by rebornix**
Implemented the manual jupyter server support through uri server provider api in `src/standalone`. Previously we had this support as a builtin concept (in other words, a special case), but now it is treated like other uri server provider (e.g. AZML).  The implementation is not perfect yet but can be used to drive the improvement for

* Jupyter server uri provider API
  * Reconnect/validation
  * Disconnect/dispose
* Kernel picker multi step workflow

This feature is behind the same setting as experimental kernel picker, so won't affect Insider users yet.

-   [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [x] Title summarizes what is changing.
-   [ ] Has a [news entry](https://github.com/Microsoft/vscode-jupyter/tree/main/news) file (remember to thank yourself!).
-   [x] Appropriate comments and documentation strings in the code.
-   [ ] Has sufficient logging.
-   [ ] Has telemetry for feature-requests.
-   [ ] Unit tests & system/integration tests are added/updated.
-   [ ] [Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.
-   [ ] [`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).

### CODE REVIEW

File: src/kernels/jupyter/jupyterUriProviderWrapper.ts [hunk 1]
```diff
@@ -43,6 +43,9 @@ export class JupyterUriProviderWrapper implements IJupyterUriProvider {
     public get displayName(): string | undefined {
         return this.provider.displayName;
     }
+    public get detail(): string | undefined {
```
Comment (rebornix) [inline src/kernels/jupyter/jupyterUriProviderWrapper.ts:46]: `id` is not sufficient for the new quick pick, thus we introduced displayname and detail. We will need to revisit the updated API soon.

File: src/kernels/jupyter/serverSelector.ts [hunk 2]
```diff
@@ -72,6 +72,57 @@ export type SelectJupyterUriCommandSource =
     | 'errorHandler'
     | 'prompt';
 
+export async function validateSelectJupyterURI(
```
Comment (rebornix) [inline src/kernels/jupyter/serverSelector.ts:75]: I thought about implementing everything completely in `standalone` but that would be quite some effort so for now I'm still using helper functions from the core.

File: src/notebooks/controllers/kernelSource/notebookKernelSourceSelector.ts [hunk 3]
```diff
@@ -77,10 +74,6 @@ interface ContributedKernelFinderQuickPickItem extends QuickPickItem {
     kernelFinderInfo: IContributedKernelFinder;
 }
 
-interface LocalJupyterServerQuickPickItem extends QuickPickItem {
```
Comment (rebornix) [inline src/notebooks/controllers/kernelSource/notebookKernelSourceSelector.ts:80]: Removed this special case in the picker and moved towards serverUriProvider api.

File: src/standalone/serviceRegistry.node.ts [hunk 4]
```diff
@@ -85,4 +86,13 @@ export function registerTypes(context: IExtensionContext, serviceManager: IServi
 
     // Dev Tools
     registerDevToolTypes(context, serviceManager, isDevMode);
+
+    const configuration = serviceManager.get<IConfigurationService>(IConfigurationService);
+    if (configuration.getSettings().kernelPickerType === 'Insiders') {
```
Comment (rebornix) [inline src/standalone/serviceRegistry.node.ts:91]: We are checking this in multiple places, wish we can have a more elegant approach ;(

File: src/standalone/userJupyterServer/userServerUrlProvider.ts [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        @inject(IConfigurationService) private readonly configService: IConfigurationService,
+        @inject(JupyterConnection) private readonly jupyterConnection: JupyterConnection,
+        @inject(IsWebExtension) private readonly isWebExtension: boolean,
+        @inject(IEncryptedStorage) private readonly encryptedStorage: IEncryptedStorage
+    ) {}
+
+    activate() {
+        this.uriProviderRegistration.registerProvider(this);
+        this._initialized = new Promise(async (resolve) => {
+            const cache = await this.encryptedStorage.retrieve(
```
Comment (rebornix) [inline src/standalone/userJupyterServer/userServerUrlProvider.ts:40]: We probably need to do a migration: copy servers manually provided from `serverUriStorage` to this tiny extension but let's play around with it first and handle the migration a bit later.

File: src/standalone/serviceRegistry.node.ts (hunk 4 above)
Comment (DonJayamanne) [reply src/standalone/serviceRegistry.node.ts:91]: I wanted to create a single class with a single readonly property,

File: src/standalone/userJupyterServer/userServerUrlProvider.ts (hunk 5 above)
Comment (DonJayamanne) [reply src/standalone/userJupyterServer/userServerUrlProvider.ts:40]: Isn't this a little risky,
Running this in `activate` will drastically slow extension loading times.
Accessing users credential store could be a very slow operation

**Review by DonJayamanne — CHANGES_REQUESTED** [review]
Lets talk about accessing data from credential store in activate event tomorrow, else everything else seems good.

**Comment by rebornix** [comment]
@DonJayamanne comments resolved, please help take another look.

File: src/standalone/userJupyterServer/userServerUrlProvider.ts [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+const UserJupyterServerUriListMementoKey = '_builtin.jupyterServerUrlProvider.uriList';
+
+@injectable()
+export class UserJupyterServerUrlProvider implements IExtensionSyncActivationService, IDisposable, IJupyterUriProvider {
+    readonly id: string = '_builtin.jupyterServerUrlProvider';
+    readonly displayName: string = DataScience.UserJupyterServerUrlProviderDisplayName();
+    readonly detail: string = DataScience.UserJupyterServerUrlProviderDetail();
+    private _onDidChangeHandles = new EventEmitter<void>();
+    onDidChangeHandles: Event<void> = this._onDidChangeHandles.event;
+    private readonly _globalDisposables: IDisposable[] = [];
```
Comment (DonJayamanne) [inline src/standalone/userJupyterServer/userServerUrlProvider.ts:44]: These aren't disposed,
you'd want to pull in `IDisposableRegistry` into the ctor

**Review by DonJayamanne — DISMISSED** [review]
Just need to dispose the disposables, else all good

**Review by DonJayamanne — APPROVED** [review]

**Pull request merged**